### DISCUSSION TITLE
How to get the block ID of the current calculating element?

### DISCUSSION THREAD

**Discussion by RECHOA** [idaholab/moose#22543] [Q&A General] [answered]
Hi all,
I want to get the block id of the current calculating element in an `Auxkernel `  type or a `Material`  type , and then use the block id as an index of a vector. 
For example, a type restricted by blocks as following:
```
  [./test]
    type = testType
    block = ' 18 23 28 33 38 43 '
  [../]
```
I tried `_current_elem->subdomain_id()` but it's incorrect.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Why do you say it's incorrect?
It should be that

Guillaume

**Reply by RECHOA** [reply]
in the source file, I writed
```
elem_subdomain_id = _current_elem->subdomain_id();
mooseError(elem_subdomain_id);
```
and test the type object  restricted by `block = ' 18 23 28 33 38 43 '` as mentioned before, but it outputs '46'.

**Reply by GiudGiud** [reply]
can you please paste your input file and which object you block restricted

**Reply by RECHOA** [reply]
the mesh and block id were generated by cubit not MOOSE, would it be the cause of the problem?

**Reply by GiudGiud** [reply]
that's unlikely to be the problem

**Reply by RECHOA** [reply]
ok I will check it again. If the problem wouldn't be solve, I'll  try to reproduce it in a simple way.

**Reply by RECHOA** [reply]
I notice that the mesh of these blocks (`block = ' 18 23 28 33 38 43 '`) are merged with block 46, which means they share a suface with block 46. so maybe when the calculting occurs in the shared suface, the element would be considered in block 46?

**Reply by GiudGiud** [reply]
Is this a material property or an auxkernel?
Computing the neighbor material property might use the neighbor subdomain, I would have to check though

**Reply by RECHOA** [reply]
this is an auxkernel.

**Reply by GiudGiud** [reply]
Then no it shouldnt happen.
Something is wrong with the block assignment or the block restriction.

**Reply by RECHOA** [reply]
I put a small case in https://github.com/RECHOA/Cases/tree/main/testblockID
In this case, the mesh of block 3 and block 8 was generated and merged together by using cubit,


![1667230528501](https://user-images.githubusercontent.com/33534979/199047378-e1520cb8-bcad-4542-b74a-359c5e54c7ad.png)

in the input file, the calculation is restricted in block 3 :
```
[AuxKernels]
  [./testid]
    type = MooseErrorSubdomainID
    variable = for_testid
    block = '3 '
  [../]
[]
```
and in the `MooseErrorSubdomainID.C` :
```
  elem_subdomain_id = _current_elem->subdomain_id();

  mooseWarning("here elem_subdomain_id ", elem_subdomain_id," .");
```
but the result shows 
`here elem_subdomain_id 8 .`

**Reply by GiudGiud** [reply]
if you use --mesh-only on the command line and look a the exodus file, do you have 2 blocks still?

**Reply by GiudGiud** [reply]
also can you please paste the command line output from that?

**Reply by RECHOA** [reply]
I run `--mesh-only`,  the exodus file still have 2 blocks and they are merged. here is the  command line output :
[command_line_output.txt](https://github.com/idaholab/moose/files/9903271/command_line_output.txt)

**Reply by hugary1995** [reply]
you need an elemental auxvar:
```
[AuxVariables]
  [for_testid]
    family = MONIMIAL
  []
[]
```

**Reply by hugary1995** [reply]
FWIW, we should probably set _current_elem to nullptr in nodal auxkernels...

**Reply by RECHOA** [reply]
@hugary1995 Yes you're right! when I add `family = MONIMIAL`, the output of the small case is reasonable now.
But acturally, I want to obtain the block id when using the moose class [RadialDisplacementSphereAux](https://mooseframework.inl.gov/source/auxkernels/RadialDisplacementSphereAux.html) , but the variable applied to this object must a nodal variable.
At the 69th row of its source file ([RadialDisplacementSphereAux.C](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/tensor_mechanics/src/auxkernels/RadialDisplacementSphereAux.C)):
```
  if (!isNodal())
    mooseError("Must run on a nodal variable");
```
so how can I get the (calculuting point?) block id?

**Reply by RECHOA** [reply]
@hugary1995  Sorry I didn't see your last reply. I will try as you mentioned:
`FWIW, we should probably set _current_elem to nullptr in nodal auxkernels...`

**Reply by RECHOA** [reply]
@hugary1995 Maybe I misunderstood what you mean. I have tried `nullptr->subdomain_id()` and `_current_node->subdomain_id()`, but they don't work.

**Reply by GiudGiud** [reply]
@RECHOA what s happening is that you are using _current_elem which is not set properly because you are looping over nodes not elements

you cannot use _current_elem, and nullptr would also not work

The nodes arent part of blocks. 
If you want block restriction, you need to use elemental variables not nodal variables

**Reply by hugary1995** [reply]
I suppose we are all on the same page now: a node may belong to multiple blocks. So you'll have to tell us how you want to handle the multiplicity... unless I misunderstood what you are trying to achieve here.

**Reply by RECHOA** [reply]
@GiudGiud The  MOOSE build-in auxkernel [RadialDisplacementSphereAux](https://mooseframework.inl.gov/source/auxkernels/RadialDisplacementSphereAux.html#) which I want to use or inherit, should be applied to nodal variable, but it could use  block restriction. The parameter 'block' is valid according to https://mooseframework.inl.gov/source/auxkernels/RadialDisplacementSphereAux.html.

@hugary1995 I want to get the radial displacements of a large number of 3D spherical objects, each object is consist of some mesh merged concentric spherical shell, and the mesh of the spherical objects are merged with another object.
Acturally, the MOOSE build-in auxkernel [RadialDisplacementSphereAux](https://mooseframework.inl.gov/source/auxkernels/RadialDisplacementSphereAux.html#) could make it in the following way:
```
  [./raddispaux_object1]
    type = RadialDisplacementSphereAux
    variable = rad_disp
    origin = '0 0 0'
    block = ‘3 4 5 6’  # block 3 4 5 are mesh merged concentric spherical shell.
  [../]
  [./raddispaux_object2]
    type = RadialDisplacementSphereAux
    variable = rad_disp
    origin = '0 100 0'
    block = ‘11 12 13 14’  # block 11 12 13 14 are mesh merged concentric spherical shell.
  [../]
......
```
but the input file would be quite long since the number of objects is large. 
So I read  all the origins of objects from a file, and I want to override the origin getting method in [RadialDisplacementSphereAux](https://mooseframework.inl.gov/source/auxkernels/RadialDisplacementSphereAux.html#) by obtaining the block id of the current calculating region, and using it as an index to get the sphere origin. 
The [RadialDisplacementSphereAux](https://mooseframework.inl.gov/source/auxkernels/RadialDisplacementSphereAux.html#) must be applied to a nodal variable, but the  parameter 'block' is valid. Now i don't know how to get the block id.

**Reply by RECHOA** [reply]
Besides, if I only apply nodal variable on boundaries, how to get the boundary id of current calculating node?

**Reply by RECHOA** [reply]
Maybe it is hardly to get correct block id of a nodal variable as an index when mesh merged. 
As a temporary method of overcoming it, now I use the element variable (strain) to calculate the radial displacement and take the block id of element variable (strain)  as the index of sphere origin.

**Reply by GiudGiud** [reply]
Hello

From the current_node, there's no efficient way to get the boundary ID or the block ID. 

There are not efficient ways (N2 scaling):
For block ID:
- from the mesh loop on elements, loop on their node, find your node and get the block ID from the element. this is not a unique block ID though, all the elements sharing that node may have different block IDs
- See below. There is a node-to-element map in the mesh you can use, unordered map lookup scaling varies

For boundary ID:
- from the mesh, loop on nodes for a given boundary ID (or for all bid, one by one), find the current node in those ranges

But do you really need to do that?
The block restriction of the variable and the kernels should have been obeyed anyway.

Guillaume

**Reply by hugary1995** [reply]
> For block ID:
> 
> from the mesh loop on elements, loop on their node, find your node and get the block ID. this is not a unique block ID though, all the elements sharing that node may have different block IDs

This could be faster by caching the node-to-element map, which moose handles for you.

But yeah, I agree with @GiudGiud: I feel you aren't using moose in the right way...

**Reply by RECHOA** [reply]
Ok it seems there is no efficient method to meet my demand for getting the boundary id or block id as an index, I will try  to solve my problem in a way that doesn't need the ids. Thanks a lot for your reply. @GiudGiud @hugary1995

**Reply by hugary1995** [reply]
I'd slightly rephrase so as not to confuse future readers of this thread: using node-to-element map to query connected elements and their block ids is fairly efficient. The map is built once and cached. The issue you were facing is the multiplicity.

When a node sits at the intersection of multiple blocks, you need to impose your own set of rules to "determine" the node's block id. To list a few possibilities:
1. You could say the smallest/largest block id of the connected element is this node's block id.
2. You could say the block id of the first connected element on the lowest rank is this node's block id.
3. You could compute the node's radial displacement based on all the connected elements' block ids, and average them.
4. ...

The list goes on. The issue is that none of them is a general purpose rule that applies everywhere. Hence, moose doesn't provide one preset rule. Users will have to use their favorite rule or apply the rule that makes the most sense for the specific use case.
